components/eam/src/physics/cam/zm_microphysics.F90

```diff
@@ -0,0 +1,3337 @@
+                 nprb(k) = prb(k)/(qcic(i,k)/ncic(i,k))
+              else
+                 prb(k)=0._r8
+                 nprb(k)=0._r8
+              end if
+              end if           
+
+              !................................................................................
+              ! heterogeneous freezing of cloud water (-5C < T < -35C)
+
+              if (qcic(i,k).ge.qsmall .and.ncic(i,k).gt.0._r8 .and. ntaerh(i,k).gt.0._r8 .and.  &
+                 t(i,k).le.268.15_r8 .and. t(i,k).gt.238.15_r8 ) then
+
+                 if (aero%scheme == 'bulk')  then
+                    ! immersion freezing (Diehl and Wurzler, 2004)
+                    ttend(k) = -grav*wu(i,k)/cp/(1.0_r8+gamhat(i,k))
+
+                    nai_bcphi = 0.0_r8
+                    nai_dst1  = 0.0_r8
+                    nai_dst2  = 0.0_r8
+                    nai_dst3  = 0.0_r8
+                    nai_dst4  = 0.0_r8
+
+                    if (aero%idxbcphi > 0) nai_bcphi = naer2h(i,k,aero%idxbcphi)
+                    if (aero%idxdst1  > 0) nai_dst1  = naer2h(i,k,aero%idxdst1)
+                    if (aero%idxdst2  > 0) nai_dst2  = naer2h(i,k,aero%idxdst2)
+                    if (aero%idxdst3  > 0) nai_dst3  = naer2h(i,k,aero%idxdst3)
+                    if (aero%idxdst4  > 0) nai_dst4  = naer2h(i,k,aero%idxdst4)
+
+                    naimm = (0.00291_r8*nai_bcphi + 32.3_r8*(nai_dst1 + nai_dst2 + &
+                       nai_dst3 + nai_dst4))/ntaerh(i,k)             !m-3
+                    if (ttend(k) .lt. 0._r8) then
+                       nnuccc(k) = -naimm*exp(273.15_r8-t(i,k))*ttend(k)*qcic(i,k)/rhow   ! kg-1s-1                        
+                       mnuccc(k) = nnuccc(k)*qcic(i,k)/ncic(i,k)
+                    end if
+                 else   
+                    if (.false.) then
+                       ! immersion freezing (Diehl and Wurzler, 2004)
+                       ttend(k) = -grav*wu(i,k)/cp/(1.0_r8+gamhat(i,k))
+                       naimm = (0.00291_r8*soot_num + 32.3_r8*dst_num )*1.0e6_r8/ntaerh(i,k)             !m-3
+                       if (ttend(k) .lt. 0._r8) then
+                          nnuccc(k) = -naimm*exp(273.15_r8-t(i,k))*ttend(k)*qcic(i,k)/rhow   ! kg-1s-1
+                          mnuccc(k) = nnuccc(k)*qcic(i,k)/ncic(i,k)
+                       end if
+                    else
+                       ! immersion freezing (Bigg, 1953)
+                       mnuccc(k) = pi*pi/36._r8*rhow* &
+                       cdist1(k)*gamma(7._r8+pgam(i,k))* &
+                       bimm*(exp(aimm*(273.15_r8-t(i,k)))-1._r8)/ &
+                       lamc(i,k)**3/lamc(i,k)**3
+
+                       nnuccc(k) = pi/6._r8*cdist1(k)*gamma(pgam(i,k)+4._r8) &
+                           *bimm*(exp(aimm*(273.15_r8-t(i,k)))-1._r8)/lamc(i,k)**3
+                    end if
+                 end if
+
+                 ! contact freezing (Young, 1974) with hooks into simulated dust
```

Comment:
@crterai not like ice nucleation for cirrus clouds, I wonder if it should be hooked with CNT heterogeneous ice nucleation scheme (hetfrz_classnuc.F90 and hetfrz_classnuc_cam.F90) or something similar to be consistent.  
CNT takes over contact freezing (Young, 1974) and heterogenous freezing for stratiform.

@hwangacme what would you think?

Reply:
@mingxuanwupnnl - I think this is something that will need to be considered at a later stage and not at this stage of the integration. This is Xiaoliang Song's response to your suggestion. 
"We discussed this with Jiwen several months ago. We think that replacing the ice nucleation scheme 
in convective microphysics at the last minute is not a very good choice. We have to rerun all the simulations and 
redo all evaluations. Since our analyses suggest that the droplet size in convective clouds is more important than 
ice particle size for ACI effect in the current model, we think we can keep the current ice nucleation scheme, and implement the CNT heterogeneous ice nucleation scheme in the future version of the model."

Reply:
@crterai I agree. We can make a PR for switching to CNT ice nucleation scheme in the future.

Reply:
This sounds good to me too.